File: packages/sdk-node/src/engine/planner.ts
/**
 * Agentic test planner - generates comprehensive test plans
 */

import { Anthropic } from "@anthropic-ai/sdk";
import { Scene, Step, Assertion, AssertionType } from "../scene/schema.js";
import { Connection } from "../connection.js";

export interface TestPlan {
  id: string;
  scene_id: string;
  name: string;
  description: string;
  test_code: string;
  test_cases: TestCase[];
  execution_order: string[];
  environment: Record<string, unknown>;
  setup_code?: string;
  teardown_code?: string;
  dependencies: string[];
  estimated_duration_ms: number;
}

export interface TestCase {
  id: string;
  name: string;
  description: string;
  steps: string[];
  assertions: string[];
  variables: Record<string, unknown>;
  timeout_ms: number;
}

const client = new Anthropic();

/**
 * Agentic planner that generates test cases and Playwright code
 */
export class AgenticPlanner {
  /**
   * Generate a comprehensive test plan from a scene
   */
  async plan(scene: Scene, connection: Connection): Promise<TestPlan> {
    const planningPrompt = this.buildPlanningPrompt(scene, connection);

    const response = await client.messages.create({
      model: "claude-opus-4-6",
      max_tokens: 4096,
      messages: [
        {
          role: "user",
          content: planningPrompt,
        },
      ],
    });

    const content = response.content[0];
    if (content.type !== "text") {
      throw new Error("Unexpected response type from Claude");
    }

    const planData = JSON.parse(content.text);

    return {
      id: `plan_${Date.now()}`,
      scene_id: scene.id,
      name: scene.title,
      description: scene.description,
      test_code: planData.test_code,
      test_cases: planData.test_cases,
      execution_order: planData.execution_order,
      environment: planData.environment || {},
      setup_code: planData.setup_code,
      teardown_code: planData.teardown_code,
      dependencies: planData.dependencies || [],
      estimated_duration_ms: planData.estimated_duration_ms || 0,
    };
  }

  /**
   * Generate actual Playwright test code from a plan
   */
  async generateTestCode(plan: TestPlan): Promise<string> {
    const codeGenPrompt = this.buildCodeGenPrompt(plan);

    const response = await client.messages.create({
      model: "claude-opus-4-6",
      max_tokens: 8192,
      messages: [
        {
          role: "user",
          content: codeGenPrompt,
        },
      ],
    });

    const content = response.content[0];
    if (content.type !== "text") {
      throw new Error("Unexpected response type from Claude");
    }

    // Extract code block if wrapped in markdown
    const text = content.text;
    const codeMatch = text.match(/```(?:typescript|javascript)?\n([\s\S]*?)\n```/);
    return codeMatch ? codeMatch[1] : text;
  }

  /**
   * Explore edge cases and generate additional test plans
   */
  async exploreEdgeCases(
    scene: Scene,
    plan: TestPlan
  ): Promise<TestPlan[]> {
    const edgeExplorationPrompt = this.buildEdgeExplorationPrompt(
      scene,
      plan
    );

    const response = await client.messages.create({
      model: "claude-opus-4-6",
      max_tokens: 6144,
      messages: [
        {
          role: "user",
          content: edgeExplorationPrompt,
        },
      ],
    });

    const content = response.content[0];
    if (content.type !== "text") {
      throw new Error("Unexpected response type from Claude");
    }

    const edgeCases = JSON.parse(content.text);
    return edgeCases.plans.map(
      (
        edgePlan: Record<string, unknown>,
        index: number
      ): TestPlan => ({
        id: `edge_${plan.id}_${index}`,
        scene_id: plan.scene_id,
        name: edgePlan.name as string,
        description: edgePlan.description as string,
        test_code: edgePlan.test_code as string,
        test_cases: (edgePlan.test_cases as TestCase[]) || [],
        execution_order: (edgePlan.execution_order as string[]) || [],
        environment: (edgePlan.environment as Record<string, unknown>) || {},
        setup_code: edgePlan.setup_code as string | undefined,
        teardown_code: edgePlan.teardown_code as string | undefined,
        dependencies: [plan.id, ...(edgePlan.dependencies as string[])],
        estimated_duration_ms:
          (edgePlan.estimated_duration_ms as number) || 0,
      })
    );
  }

  /**
   * Build the planning prompt for Claude
   */
  private buildPlanningPrompt(scene: Scene, connection: Connection): string {
    return `You are an expert QA engineer. Generate a comprehensive test plan for the following scenario.

SCENARIO:
Title: ${scene.title}
Description: ${scene.description}
Priority: ${scene.priority}
Actor: ${scene.actor ? JSON.stringify(scene.actor) : "Not specified"}

STEPS:
${scene.steps.map((s, i) => `${i + 1}. ${s.action} ${s.input ? `with ${JSON.stringify(s.input)}` : ""}`).join("\n")}

ASSERTIONS:
${scene.assertions.map((a) => `- ${a.type} on ${a.target}: expect ${JSON.stringify(a.expected)}`).join("\n")}

CONNECTION:
Type: ${connection.type}
Base URL: ${connection.base_url}
Timeout: ${connection.timeout}ms

Return a JSON object with the following structure:
{
  "test_cases": [
    {
      "id": "test_1",
      "name": "Test case name",
      "description": "What it tests",
      "steps": ["Step 1", "Step 2"],
      "assertions": ["Assert X", "Assert Y"],
      "variables": {},
      "timeout_ms": 30000
    }
  ],
  "execution_order": ["test_1", "test_2"],
  "environment": {},
  "setup_code": "// setup steps",
  "teardown_code": "// cleanup",
  "dependencies": [],
  "estimated_duration_ms": 30000,
  "test_code": "// Playwright test code will be generated separately"
}

Generate realistic, executable test cases that cover the scenario.`;
  }

  /**
   * Build the code generation prompt
   */
  private buildCodeGenPrompt(plan: TestPlan): string {
    return `Generate production-quality Playwright test code for the following test plan:

Test Plan Name: ${plan.name}
Description: ${plan.description}

Test Cases:
${plan.test_cases.map((tc) => `- ${tc.name}: ${tc.description}`).join("\n")}

Requirements:
1. Use Playwright with TypeScript
2. Import from '@playwright/test'
3. Use test.describe() for grouping
4. Include proper error handling
5. Use assertions from expect()
6. Add meaningful comments
7. Handle async/await correctly
8. Include timeout configuration
9. Support the environment variables: ${Object.keys(plan.environment).join(", ")}

Return ONLY the complete, executable Playwright test code in a TypeScript code block.`;
  }

  /**
   * Build the edge case exploration prompt
   */
  private buildEdgeExplorationPrompt(scene: Scene, plan: TestPlan): string {
    return `As a QA expert, identify edge cases and boundary conditions for this scenario and generate additional test plans.

Original Scenario: ${scene.title}
${scene.description}

Current Test Plan Covers:
${plan.test_cases.map((tc) => `- ${tc.name}`).join("\n")}

Suggested Edge Cases from Scenario:
${scene.edge_cases.map((ec) => `- ${ec}`).join("\n")}

Generate test plans for these categories:
1. Boundary values (min, max, zero, negative)
2. Authentication/authorization edge cases
3. Timing and race conditions
4. Data validation edge cases
5. Network/connectivity failures
6. Concurrent operations
7. Invalid inputs and malformed data
8. State transition edge cases

Return a JSON object with structure:
{
  "plans": [
    {
      "name": "Edge case name",
      "description": "What the edge case tests",
      "test_code": "Playwright code",
      "test_cases": [],
      "execution_order": [],
      "environment": {},
      "dependencies": [],
      "estimated_duration_ms": 30000
    }
  ]
}`;
  }
}

File: packages/sdk-python/rigour_qa/engine/planner.py
"""Agentic planner for generating test plans from scenes."""

import json
from typing import Optional
from uuid import uuid4

from anthropic import Anthropic

from rigour_qa.connection import Connection
from rigour_qa.engine.types import TestPlan
from rigour_qa.scene.schema import Scene


class AgenticPlanner:
    """Plan generator that uses Claude to understand scenes and generate executable test code."""

    def __init__(self, api_key: Optional[str] = None):
        """Initialize planner with optional Anthropic API key."""
        self.client = Anthropic(api_key=api_key)

    def plan(self, scene: Scene, connection: Connection) -> TestPlan:
        """Generate a test plan from a scene."""
        prompt = self._build_planning_prompt(scene, connection)

        message = self.client.messages.create(
            model="claude-opus-4-6",
            max_tokens=4096,
            messages=[{"role": "user", "content": prompt}]
        )

        response_text = message.content[0].text
        test_code = self._extract_code(response_text)

        if not test_code:
            test_code = self._generate_fallback_code(scene, connection)

        plan = TestPlan(
            id=str(uuid4()),
            scene_id=scene.id,
            title=scene.title,
            description=scene.description,
            test_code=test_code,
            steps_count=len(scene.steps) if scene.steps else 1,
        )

        return plan

    def _build_planning_prompt(self, scene: Scene, connection: Connection) -> str:
        """Build the prompt for test code generation."""
        steps_desc = ""
        if scene.steps:
            steps_desc = "Steps:\n"
            for i, step in enumerate(scene.steps, 1):
                steps_desc += f"  {i}. {step.action}"
                if step.expect:
                    steps_desc += f" (expect: {step.expect})"
                steps_desc += "\n"

        assertions_desc = ""
        if scene.assertions:
            assertions_desc = "Assertions:\n"
            for assertion in scene.assertions:
                assertions_desc += f"  - {assertion.type}: {assertion.target} == {assertion.expected}\n"

        return f"""Generate a Python pytest test function for this test scenario.

Test Title: {scene.title}
Description: {scene.description}
Base URL: {connection.base_url}
{steps_desc}
{assertions_desc}

Requirements:
1. Use httpx for HTTP requests
2. Include all assertions from the description
3. Use pytest conventions (assert statements)
4. Handle both success and error cases
5. Make it production-quality, not a stub
6. Name function: test_{scene.id.replace('-', '_')[:20]}
7. Return bare Python code (no markdown, no explanation)
8. Include proper error handling and logging

Generate the test function now:"""

    def generate_test_code(self, plan: TestPlan) -> str:
        """Return the test code from a plan."""
        return plan.test_code

    def explore_edge_cases(
        self, scene: Scene, plan: TestPlan, connection: Connection
    ) -> list[TestPlan]:
        """Autonomously generate edge case variants of a test plan."""
        prompt = f"""Given this test scenario:
Title: {scene.title}
Description: {scene.description}

Generate 5 edge case variations of this test. For each, provide:
1. Edge case name
2. What makes it an edge case
3. Expected behavior

Focus on:
- Boundary values (empty, max length, special characters)
- Authentication edge cases (invalid token, expired, wrong role)
- Timing edge cases (concurrent requests, slow responses)
- Data edge cases (null fields, duplicates, unicode)
- Network edge cases (timeouts, connection errors)

Return JSON array: [{{"name": "...", "description": "...", "focus": "..."}}]"""

        message = self.client.messages.create(
            model="claude-opus-4-6",
            max_tokens=2048,
            messages=[{"role": "user", "content": prompt}]
        )

        response_text = message.content[0].text

        try:
            import re
            json_match = re.search(r'\[.*\]', response_text, re.DOTALL)
            if json_match:
                edge_cases_data = json.loads(json_match.group())
            else:
                edge_cases_data = json.loads(response_text)
        except json.JSONDecodeError:
            edge_cases_data = []

        plans = []
        for i, edge_case in enumerate(edge_cases_data[:5]):
            edge_scene = Scene(
                id=str(uuid4()),
                title=f"{scene.title} - {edge_case.get('name', f'Edge Case {i+1}')}",
                description=f"{scene.description}\nEdge case: {edge_case.get('description', '')}",
                actor=scene.actor,
                tags=(scene.tags or []) + ["edge-case"],
                priority=scene.priority,
            )

            edge_plan = self.plan(edge_scene, connection)
            plans.append(edge_plan)

        return plans

    def _extract_code(self, response: str) -> Optional[str]:
        """Extract Python code from response."""
        import re
        code_match = re.search(
            r'```python\n(.*?)\n```',
            response,
            re.DOTALL
        )
        if code_match:
            return code_match.group(1)

        if response.strip().startswith("def test_"):
            return response.strip()

        return None

    def _generate_fallback_code(self, scene: Scene, connection: Connection) -> str:
        """Generate a fallback test when Claude fails - creates actual HTTP calls."""
        test_name = scene.id.replace('-', '_')[:20]

        # Build step descriptions
        steps_code = ""
        if scene.steps:
            for i, step in enumerate(scene.steps, 1):
                # Parse action as HTTP method and path
                parts = step.action.split(' ', 1)
                method = parts[0] if len(parts) > 0 else "GET"
                path = parts[1] if len(parts) > 1 else "/"

                steps_code += f"\n    # Step {i}: {step.action}\n"
                if method.upper() in ["GET", "DELETE", "HEAD"]:
                    steps_code += f'    response = client.{method.lower()}("{path}")\n'
                else:
                    json_str = json.dumps(step.input or {})
                    steps_code += f'    response = client.{method.lower()}("{path}", json={json_str})\n'

                if step.expect:
                    steps_code += f'    # Expected: {step.expect}\n'

        # Build assertions
        assertions_code = ""
        if scene.assertions:
            for assertion in scene.assertions:
                if assertion.type.value == "status_code":
                    assertions_code += f'\n    assert response.status_code == {assertion.expected}, "Status code should be {assertion.expected}"\n'
                elif assertion.type.value == "body_contains":
                    assertions_code += f'\n    assert "{assertion.expected}" in response.text, "Response should contain {assertion.expected}"\n'
        else:
            # Default: assert not a 5xx error
            assertions_code = '\n    assert response.status_code < 500, "Should not return server error"'

        code = f'''import httpx
import pytest

def test_{test_name}():
    """Test: {scene.title}
    {scene.description}
    """
    client = httpx.Client(base_url="{connection.base_url}")
{steps_code}
{assertions_code}
'''
        return code
